### ISSUE TITLE
Using PrimeVue within a vue component library

### ISSUE THREAD

**Issue by johnwc** [primefaces/primevue#6678]
### Describe the bug

We are using PrimeVue in one of our vue component libraries and then importing the library within our main web application. The theme does not seem to be added to the application's styling. Using DataTable as an example, the table is rendering and working as expected. But it is not styled at all. If we manually add a DataTable component within the same page that is importing our component from our library, then the theme gets injected into the page and both the DataTable we added as well as the DataTable being used in the library are styled correctly. It seems that PrimeVue config is not writing the css to the page when it doesn't think the component is used, not realizing it is being used from imported library.

There does not seem to be a way to have PrimeVue config write out the css for components it doesn't know about.

### Reproducer

N/A

### PrimeVue version

4.0.7

### Vue version

4.x

### Language

TypeScript

### Build / Runtime

Vite

### Browser(s)

Any

### Steps to reproduce the behavior

1. Create vue component library
2. Use PrimeVue component(DataTable) within a custom component in library.
3. Build & publish library to npm repo.
4. Create vue web app
5. Import your library from npm repo
6. Configure web app with PrimeVue config, using Aura theme
  ```typescript
const app = createApp(App)
    .use(PrimeVue, {
        theme: {
            preset: Aura,
            options: {
                prefix: 'p',
                darkModeSelector: '.app-dark',
                cssLayer: false
            }
        }
    })

```
7. Use custom component from library in page; seeing it contain no style.
8. Add same PrimeVue type component(DataTable), that was used in custom component, above/below custom component; see both have style applied.

### Expected behavior

Adding PrimeVue config with theme configured produces css for any component used in page or library components.
 ```typescript
const app = createApp(App)
    .use(PrimeVue, {
        theme: {
            preset: Aura,
            options: {
                prefix: 'p',
                darkModeSelector: '.app-dark',
                cssLayer: false
            }
        }
    })

```

**Comment by roelhem** [comment]
I've stumbled on this problem to today.
For me, the problem was the ui library had `primevue` as a normal dependency in the package.json. This resulted in two different PrimeVue versions in the consumer package... The used theme is registered on a global `Theme` object that does not use the provide/inject mechanism of Vue. In my situation, I had two different `Theme` objects: one to which the ui libraries were listening and a different one on which the theme was actually set.

I solved my case by making `primevue` a peer dependency in the ui library to ensure that I had only one version of `primevue` in the consuming package.

**Comment by johnwc** [comment]
> I've stumbled on this problem to today. For me, the problem was the ui library had `primevue` as a normal dependency in the package.json. This resulted in two different PrimeVue versions in the consumer package... The used theme is registered on a global `Theme` object that does not use the provide/inject mechanism of Vue. In my situation, I had two different `Theme` objects: one to which the ui libraries were listening and a different one on which the theme was actually set.
> 
> I solved my case by making `primevue` a peer dependency in the ui library to ensure that I had only one version of `primevue` in the consuming package.

Can you give an example of what you are referring to as in code examples, and what you are calling in the library to get the theme to even work in the primary app. I believe you are getting it to work because you are using primevue components in the primary app, and the plugin detects them and injects the css in the page which everything uses. Try moving all primevue usage to the library, then apply the theme in the primary app. This is where we are not seeing any css being injected into the page, as the PrimeVue plugin is not detecting any PrimveVue components in the app itself.

**Comment by cristinafernandez1234** [comment]
Hi! I don't know if the issue you have is the same as mine, I explained it in the discussions page. I didn't find the solution yet, although other person did it for their project: https://github.com/orgs/primefaces/discussions/2436. 

You can inspect the elements with attribute "data-primevue-style-id" both in the library and in the main web application. In the library you would have both  `<style type="text/css" data-primevue-style-id="datatable-style">`  and  `<style type="text/css" data-primevue-style-id="datatable-variables">` and in the main web application only `<style type="text/css" data-primevue-style-id="datatable-style">` , the component variables are missing.

**Comment by johnwc** [comment]
@cristinafernandez1234 yes, you are seeing the same issue as we are. And you are doing the same manual process to work around it, by copying the produced css out to a global style sheet.

**Comment by hisuwh** [comment]
@johnwc have you made any progress on this.  We're seeing the same issue

**Comment by hisuwh** [comment]
Seeing some weird behaviour with this.  The workaround I've seen suggested is to import the primevue components directly.  If I import and render the `Menu` component then our component using Menu displays correctly.

However, I have another component using `Card` and importing and using `Card` directly doesn't fix the styles.

This feels like too much magic.  Can there not be a way for us to just explicitly import the styles we need?

**Comment by johnwc** [comment]
@hisuwh no, the only thing I was able to do was to render all components in a test page to allow me to copy the css injected into the page. I then take the copy and save it into css files to manually import on the pages we use the components on. And, I've received no response from anyone other than others having the same issue.

**Comment by johnwc** [comment]
@ArthurN instead of asking a bunch of questions about your primevue usage in each project, can you create a sample repo with it working?

**Comment by cristinafernandez1234** [comment]
Hi @ArthurN and @johnwc, I tried the solution from Arthur and it worked for me.

In my case it was only needed to add this in peerDependencies:

```
"peerDependencies": {
    "vue": "^3.2.31",
    "@primeuix/styled": "^0.0.5",
    "@primevue/themes": "^0.0.5",
    "primevue": "^4.2.2"
  }
```

**Comment by hisuwh** [comment]
I had to do this to get the styles for the referenced imported components working:
```
# PrimeShim.vue
<script lang="ts" setup>
import * as PrimeVue from "primevue";

const items: any[] = [
    "Button",
    "Card",
    "Menu"
];
</script>

<template>
    <div :class="style['prime-shim']">
        <component
            :is="(PrimeVue as any)[item]"
            v-for="item of items"
            :key="item"
        />
    </div>
</template>
<style module="style">
.prime-shim {
    display: none;
}
</style>

```

**Comment by erwanlfrt** [comment]
I was facing a similar issue and after analysing the bundle of my app, I noticed 2 objects `Theme` instead of one like @roelhem described in [his comment](https://github.com/primefaces/primevue/issues/6678#issuecomment-2447953450). leading to missing theme injection.
In my case, I'm not installing my custom component library using npm but I'm resolving it locally instead so I'm not sure that my solution will work for you. In my case, using vite config [resolve.dedupe](https://vite.dev/config/shared-options#resolve-dedupe) fixed my issue.

- `vite.config.js` of my component library:
```js
export default defineConfig({
    build: {
        ...
        rollupOptions: {
            external: [
	        'primeicons',
		/^primevue(\/.+)?$/,
            ]
       }
  }
});
```

- `vite.config.js` of my app:
```js

export default defineConfig({
    resolve: {
        dedupe: [
	    'primevue'
        ],
    }
 })
```

I hope it helps.

**Comment by johnwc** [comment]
> Hi [@ArthurN](https://github.com/ArthurN) and [@johnwc](https://github.com/johnwc), I tried the solution from Arthur and it worked for me.
> 
> In my case it was only needed to add this in peerDependencies:
> 
> ```
> "peerDependencies": {
>     "vue": "^3.2.31",
>     "@primeuix/styled": "^0.0.5",
>     "@primevue/themes": "^0.0.5",
>     "primevue": "^4.2.2"
>   }
> ```

So, this has gotten it to work, but when trying to migrate to the newer `primeuix/themes` package, it no longer works.

**Comment by Maksk89** [comment]
Hello. I'm having the same issue using my component library:
package.json of a **component library**
 ```
"peerDependencies": {
    "vue": "^3.4.27",
    "primevue": "^4.3.3",
    "@primeuix/themes": "^1.0.1",
    "primeflex": "^4.0.0",
    "primeicons": "^7.0.0"
  }
```
Primevue styles do not apply to my main app in this case.
Can you help me to fix it, please? Thank you

**Comment by cristinafernandez1234** [comment]
What we do is exporting the theme (example CustomPreset) from the component library and use it in the project:

import {
  PrimeVue,
  FlagshipPreset
} from "component-library";

 app.use(PrimeVue, {
    theme: {
      preset: CustomPreset,
      options: {
        darkModeSelector: false
      }
    }
  });

**Comment by johnwc** [comment]
I found that every custom library that uses primevue, must export the Config type from primevue. The in the main app, it must call each using the app.use(...).

I think this has something to do with they are using symbols for the theme object, and symbols get applied during build I believe.

**Comment by Maksk89** [comment]
> I found that every custom library that uses primevue, must export the Config type from primevue. The in the main app, it must call each using the app.use(...).
> 
> I think this has something to do with they are using symbols for the theme object, and symbols get applied during build I believe.

Could you explain this? How should I act to  export the Config type from primevue (in my component library)?

**Comment by johnwc** [comment]
@Maksk89 you can find a better explanation in my other issue here #7621 , where I pointed out more in detail the underlining issue. But, they closed it as saying they needed a test project to view the problem, even though I stated I could create a sample repo if needed. Guess they just read the first paragraph and move on to other issues without reading entire content.

**Comment by johnwc** [comment]
@Maksk89 you need to create a new issue and ask it there. I don't think this is related to this issue.

**Comment by Maksk89** [comment]
Thank you

**Comment by Maksk89** [comment]
Could you help me with another question, please?
In my library I have 
**primevue.ts**
```
import type { App } from "vue";

import PrimeVue from "primevue/config";
import Button from "primevue/button";
import Toast from "primevue/toast";
import ToastService from "primevue/toastservice";
import Ripple from "primevue/ripple";
import Lara from "@primeuix/themes/lara";
import { definePreset } from "@primeuix/themes";
import "../styles/fabrix-colors.css";
import { palette } from "@primeuix/themes";

export const applyThemeColors = () => {
const computedStyle = getComputedStyle(document.documentElement);
const primaryColorValue = computedStyle.getPropertyValue("--primary-color");
const primaryColor = palette(primaryColorValue);

return {
primary: primaryColor,
secondary: {
50: "{gray.50}",
100: "{gray.100}",
200: "{gray.200}",
300: "{gray.300}",
400: "{gray.400}",
500: "{gray.500}",
600: "{gray.600}",
700: "{gray.700}",
800: "{gray.800}",
900: "{gray.900}",
950: "{gray.950}",
},
};
};

export const MyPreset = definePreset(Lara, {
semantic: applyThemeColors(),
//Your customizations, see the following sections for examples
});

export const PrimeVuePlugin = {
install: (app: App) => {
applyThemeColors();
app.use(PrimeVue, {
ripple: false,
theme: {
preset: MyPreset,
options: {
prefix: "p",
darkModeSelector: "system",
cssLayer: {
name: "primevue",
order: "prime, primevue, fabrix",
},
},
},
});
app.use(ToastService);
app.component("Button", Button);
app.component("Toast", Toast);
app.directive("ripple", Ripple);
},
};
```
I use this library in another like this:
**main.js**
```
import "primeicons/primeicons.css";
import { createApp } from "vue";
import App from "./App.vue";
import router from "./router";
// Fabrix styling:
import { UiKitPlugin, MyPreset } from "@fabrix/ui-kit";
import { applyThemeColors } from "@fabrix/ui-kit";
import "@fabrix/ui-kit/ui-kit.css";
import PrimeVue from "primevue/config";

App.created = function () {
if (this.created) this.created();
this.$auth.init();
};
const app = createApp(App);
console.log("First version of preset is", MyPreset);

const themeColors = applyThemeColors();
console.log("Theme colors", themeColors);

const themeConfig = {
...MyPreset,
semantic: {
...MyPreset.semantic,
primary: themeColors.primary,
secondary: themeColors.secondary,
},
};

console.log("Updated preset is", themeConfig);

app.use(PrimeVue, {
theme: { preset: themeConfig },
});

app.use(UiKitPlugin, import.meta.env);
app.mount("#app");
export default app;
```
The colors from the library are not being applied in the project that uses it.

Though the colors in the **themeConfig** object are exactly as expected and correct, but they are not being applied despite using the following configuration:
```
app.use(PrimeVue, {
  theme: { preset: themeConfig },
});
```
How should I act to fix this?

**Comment by johnwc** [comment]
@Maksk89 you need to ask your question in a new issue, this has nothing to do with this issue.

**Comment by ac-zmcnulty** [comment]
Does anyone have a grasp on which package(s) is responsible for injecting component styles into the apps that need them? I'm running into a similar issue, but in the context of running an app in a shadow dom -- still, i've got the same symptoms you describe: no component styling in nested components unless the nested component is rendered in some file higher up the directory tree.

**Comment by johnwc** [comment]
> Does anyone have a grasp on which package(s) is responsible for injecting component styles into the apps that need them? I'm running into a similar issue, but in the context of running an app in a shadow dom -- still, i've got the same symptoms you describe: no component styling in nested components unless the nested component is rendered in some file higher up the directory tree.

See my [comment](https://github.com/primefaces/primevue/issues/6678#issuecomment-2843220699) above, you'll need to export each and register them all in the main app. It's a mess to be honest, this is one of a few reasons we are slowly migrating to another UI lib. This dynamic injection of styles is causing to much time to be waisted.

**Comment by digitalextremist** [comment]
@ac-zmcnulty / @johnwc I am seeing this exact same issue and trying to work around it; well described there.

I too need to decide to somehow untangle this mess of `PrimeVue` code, or switch to another UI framework, except in my case, cannot do it incrementally. Have also lost a _lot_ of time on this to the point it is untenable to stick with `PrimeVue`

If I find any solution I will fire back on this, but otherwise it seems there is a collective dead-end so far. Feels like washing a floor with muddy water at this point.

I went as far as to use the @hisuwh `shim` approach, except instantiating one of _every type of component_ to see if that would get the injections to fire and style the application properly, but that also did not work: https://github.com/primefaces/primevue/issues/6678#issuecomment-2531555477

---

I have even gone as far to completely remove the 'external' package and directly include the library within the same project as the end-result application, and still no success.

The strange thing is, I have another application ( a test application in the original library ) and it renders perfectly.

Here is the key difference between the functioning and the not-functioning application which are nearly identical:

## Functional, inspecting a `DynamicDialog` component that is open:

<img width="509" height="812" alt="Image" src="https://github.com/user-attachments/assets/e176f260-978d-4960-ab0d-0f95b22a26ae" />

## Dysfunctional, inspecting the same `DynamicDialog` component:

<img width="499" height="490" alt="Image" src="https://github.com/user-attachments/assets/4fd4ade5-9b76-4d4b-a334-d50a7c584856" />

Note, no `layer: primevue` in the dysfunctional one, and no colors being set. All those values are grayed out.

And then suddenly `InputText` gets activated somehow, in a subsequent component:

<img width="494" height="366" alt="Image" src="https://github.com/user-attachments/assets/0d306707-5a83-4bc8-b9f1-85ae01fd583e" />

Notice that is also not in any particular layer.

And this is what the styling was before a nested component somehow changed the entire definition of `InputText`

<img width="492" height="328" alt="Image" src="https://github.com/user-attachments/assets/d2a6cfab-2d54-4237-91f4-7aca79571f3e" />

As @hisuwh said, "feels like too much magic" ... and tracing through this repo versus that repo and this package versus that package ... some which are just references to other packages ... feels like taking a tour of technical debt.

Not sure if there actually can be a solution here other than a temporary patch, since the overall library seems to be in disrepair by over-engineering... and then not pruned down to what works. It seemed like it had tremendous potential but if you go beyond just basic usage, it starts to break down pretty fast.

The only possible issue I can see remaining which could be my fault, or something I can change, is `package.json` issues as mentioned by @cristinafernandez1234, which I am trying... and possible `vite.config.js` issues... but there are so many different versions and incarnations of packages it will take time to unravel... not boding well!

**Comment by digitalextremist** [comment]
Woke up with a possible solution I am going to try: In some discussions on `discord` this issue was talked through some more, and I have tried solutions given here. All roads point to `Vite` not being able to construe all the levels of abstraction mixed with package redirection mixed with `Vue` metaprogramming in `PrimeVue` and my own libraries and applications.

The dynamic I have not fully factored in is `Vite` build aliases, which in my case go across project directory boundaries. Will report back. Either way it seems `PrimeVue` is due for refactoring to simplify the level of abstraction and metaprogramming.

**Comment by digitalextremist** [comment]
Initial attempts at `Vite` optimizations did not work so I am going to thread-off into this issue: https://github.com/primefaces/primevue/issues/8126

If I find a fix I will come back to this issue and request validation of that fix and see if it also resolves what is being seen here. I _believe_ we are all experiencing the same issue but am going to drop that assumption until proven.

**Comment by digitalextremist** [comment]
_Almost_ found a solution that does not require a `shim` ... have verified it is a bundling issue.

The give-away is that it occurs in "vue component library" which actually means "code at a non-standard path" ...

`PrimeVue` abstraction/metaprogramming/obfuscation notwithstanding, `Vite` being overwhelmed is solvable, but knowing what `optimizeDeps` setting is difficult. I have it resolved, but only in `dev` ... `build` still requires a shim.

Here is my approach to try and see if that fixes it and removes the `'magic'` feeling at the same time... but only in `dev` until someone hopefully comes through with the rest of the solution:

https://github.com/primefaces/primevue/issues/8126#issuecomment-3318185733

And this is the updated version of the @hisuwh solution which _removes_ the components also:

```vue
<script lang="ts" setup>
import { ref, onMounted } from "vue";
import * as PrimeVue from "primevue";

const items = ref<string[]>([
  "Button",
  "Card",
  "Dialog",
  "InputText",
  "Toolbar",
  "Panel",
  "Toast",
  "Textarea",
  "FloatLabel",
  //de ... add all the components you use ...
]);
onMounted(() => {
  setTimeout(() => {
    items.value = [];
  }, 1000); //de Set the timer to whatever you prefer; this is 1 second.
});
</script>

<template>
  <div :class="style['shim']">
    <template v-for="key of items">
      <component :is="(PrimeVue as any)[key]" />
    </template>
  </div>
</template>
<style module="style">
.shim {
  display: none;
}
</style>
```

... otherwise you will see double with `Toast` and have phantom components.

**Comment by digitalextremist** [comment]
Just noticed the `shim` approach adds `~2mb` to package size? 😬

---

Updated version of `shim` does not bundle the _entire_ `PrimeVue` library, but selects the components:

```vue
<script lang="ts" setup>
import { ref, onMounted } from "vue";
import { Button, Toast } from "primevue";
const PrimeVue = { Button, Toast };

const items = ref<string[]>([
  "Button",
  "Toast",
]);
onMounted(() => {
  setTimeout(() => {
    items.value = [];
  }, 1000);
});
</script>

<template>
  <div :class="style['shim']">
    <template v-for="key of items">
      <component :is="(PrimeVue as any)[key]" />
    </template>
  </div>
</template>
<style module="style">
.shim {
  display: none;
}
</style>
```
